### ISSUE TITLE
IntelliJ doesn't complete "if"

### ISSUE THREAD

**Issue by feinstein** [flutter/flutter-intellij#3279]
When I type `if` in IntelliJ IDEA for Java, it autocompletes to `if ([caret-here])`, but this isn't working in Dart.

**Comment by kenzieschmoll** [comment]
@jwren @lambdabaa thoughts?

**Comment by stevemessick** [comment]
This doesn't work for kotlin, either. @alexander-doroshko what's this called? Is it just normal code completion or something special like a template? I checked youtrack.jetbrains.com but couldn't find an issue for this. In any case, this is a Dart issue, not Flutter.

**Comment by jwren** [comment]
This is the live template support.  See https://github.com/JetBrains/intellij-plugins/blob/14ede6746b54e15c6521619b8e5a551a1d6f2dfb/Dart/resources/liveTemplates/dart_miscellaneous.xml

In general we aim to match java support, search for `sout` and `soutm` in the above link.  This approach appears to be correct given the opening comment.  I'll create a PR in the coming days.

**Comment by alexander-doroshko** [comment]
I haven't found 'if' live template in Java support. Looks like it's one of Java CompletionContributors. And it behaves smarter than live template would behave. It shows up only where 'if' statement is available. Analysis Server might want to add this. However 'if' word is so short that such completion item would save no more than one keystroke (provided that closing parenthesis is auto-inserted).

**Comment by feinstein** [comment]
Well, you already have `sout` that autocompletes to `print()`, which is just 1 character difference. 

Besides the `if ()` style has a space between the `if` and the `()` so that's one more character.

Not even mentioning that after completing the sentence to be evaluated inside the, `if`, the caret will be placed inside the curly braces on another line.

**Comment by alexander-doroshko** [comment]
> you already have `sout`

Right, that's for Java developers who are used to `sout` in Java where it expands to `System.out.println()` :)

> a space between the if and the () so that's one more character

I've counted that. `if<Tab>` vs `if<Space>(`.

> the caret will be placed inside the curly braces on another line

Hm, `if<Tab>` doesn't add curly braces for me. Only `if (<caret>)`.

> it's a nice thing to have

I don't mind in general. Just wanted to discuss whether this should be a live template or Analysis Server completion contribution. I'm inclined to the latter.
BTW looks like Analysis Server already contributes 'if' item, but without `()` on selection. As far as I remember Server may tune this item so that IDE inserts `if ()`, and no action on Dart plugin end required. But I'm not 100% sure.

**Comment by feinstein** [comment]
>I've counted that. if<Tab> vs if<Space>(.

Well, actually is `i<Tab>` because I get an autocomplete suggestion for `if` as soon as I press `i`.

>Hm, if<Tab> doesn't add curly braces for me. Only if (<caret>).

You are right, I was speaking from memory, I have to do an Enter at the end of the `)` to get the curly braces, IIRC. 

I am not experienced enough to give an opinion about the best way to implement this unfortunately.

**Comment by stevemessick** [comment]
> BTW looks like Analysis Server already contributes 'if' item, but without () on selection. 

Interestingly, if you look at the completion list you'll see "if" with no parens (in Java) but the parens are added when you choose the if-item from the list. Not sure if that makes a difference to Dart, but it is a curiosity.

**Comment by feinstein** [comment]
I have seen that there's an autocomplete to `if` I just don't see the point of it, just 1 character difference after all. A change to include the `()` should be welcome in this case.

**Comment by alexander-doroshko** [comment]
> you'll see "if" with no parens (in Java) but the parens are added when you choose the if-item from the list

No worries, Analysis Server API already allows to show one text and to insert another one and also to control final caret position.
